plugins/outputs/azuremonitor/azuremonitor.go

```diff
@@ -0,0 +1,445 @@
+package azuremonitor
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"hash/fnv"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"regexp"
+	"time"
+
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/metric"
+	"github.com/influxdata/telegraf/plugins/outputs"
+)
+
+var _ telegraf.AggregatingOutput = (*AzureMonitor)(nil)
+var _ telegraf.Output = (*AzureMonitor)(nil)
+
+// AzureMonitor allows publishing of metrics to the Azure Monitor custom metrics service
+type AzureMonitor struct {
+	useMsi              bool              `toml:"use_managed_service_identity"`
+	ResourceID          string            `toml:"resource_id"`
+	Region              string            `toml:"region"`
+	Timeout             internal.Duration `toml:"Timeout"`
+	AzureSubscriptionID string            `toml:"azure_subscription"`
+	AzureTenantID       string            `toml:"azure_tenant"`
+	AzureClientID       string            `toml:"azure_client_id"`
+	AzureClientSecret   string            `toml:"azure_client_secret"`
+	StringAsDimension   bool              `toml:"string_as_dimension"`
+
+	url         string
+	msiToken    *msiToken
+	oauthConfig *adal.OAuthConfig
+	adalToken   adal.OAuthTokenProvider
+
+	client *http.Client
+
+	cache map[time.Time]map[uint64]*aggregate
+}
+
+type aggregate struct {
+	telegraf.Metric
+	updated bool
+}
+
+const (
+	defaultRegion          string = "eastus"
```

Comment:
We shouldn't have a default region constant.  This value either needs to come from instance metadata or come from user configuration.  The region must match the region of the Azure resource ID and can't be guessed.